### PR TITLE
Convert ha-cover-controls to TypeScript/LitElement

### PR DESCRIPTION
**Pull request by iantrich**
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-cover-controls.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  property,
+} from "lit-element";
+import { classMap } from "lit-html/directives/class-map";
+
+import { UNAVAILABLE } from "../data/entity";
+import { HomeAssistant } from "../types";
+import { HassEntity } from "home-assistant-js-websocket";
+import CoverEntity from "../util/cover-model";
+
+import "./ha-icon-button";
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:16]:
```suggestion
import {
  css,
  CSSResult,
  customElement,
  html,
  LitElement,
  property,
} from "lit-element";
import { classMap } from "lit-html/directives/class-map";
import { HassEntity } from "home-assistant-js-websocket";

import { UNAVAILABLE } from "../data/entity";
import type { HomeAssistant } from "../types";
import CoverEntity from "../util/cover-model";

import "./ha-icon-button";
```

File: src/components/ha-cover-controls.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  @property() public stateObj!: HassEntity;
+
+  protected render() {
+    const entityObj = new CoverEntity(this.hass, this.stateObj);
+
+    return html`
+      <div class="state">
+        <ha-icon-button
+          class=${classMap({
+            invisible: entityObj.supportsOpen,
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:31]: Can't we just not render the icon button at all?

```ts
${entityObj.supportsOpen ? html`` : ""}
```

File: src/components/ha-cover-controls.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+      .state {
+        white-space: nowrap;
+      }
+      [invisible] {
+        visibility: hidden !important;
+      }
+    `;
+  }
+
+  private computeOpenIcon(): string {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:71]:
```suggestion
  private _computeOpenIcon(): string {
```

File: src/components/ha-cover-controls.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      case "awning":
+      case "door":
+      case "gate":
+        return "hass:arrow-expand-horizontal";
+      default:
+        return "hass:arrow-up";
+    }
+  }
+
+  private computeCloseIcon(): string {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:82]:
```suggestion
  private _computeCloseIcon(): string {
```

File: src/components/ha-cover-controls.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        return "hass:arrow-down";
+    }
+  }
+
+  private computeStopDisabled(): boolean {
+    if (this.stateObj.state === UNAVAILABLE) {
+      return true;
+    }
+    return false;
+  }
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:98]: This one can probably be done inline?
```suggestion
  private _computeStopDisabled(): boolean {
    if (this.stateObj.state === UNAVAILABLE) {
      return true;
    }
    return false;
  }
```

File: src/components/ha-cover-controls.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+  private computeOpenDisabled(): boolean {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:100]:
```suggestion
  private _computeOpenDisabled(): boolean {
```

File: src/components/ha-cover-controls.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    if (this.stateObj.state === UNAVAILABLE) {
+      return true;
+    }
+    const assumedState = this.stateObj.attributes.assumed_state === true;
+    const entityObj = new CoverEntity(this.hass, this.stateObj);
+    return (entityObj.isFullyOpen || entityObj.isOpening) && !assumedState;
+  }
+
+  private computeClosedDisabled(): boolean {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:109]:
```suggestion
  private _computeClosedDisabled(): boolean {
```

File: src/components/ha-cover-controls.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    if (this.stateObj.state === UNAVAILABLE) {
+      return true;
+    }
+    const assumedState = this.stateObj.attributes.assumed_state === true;
+    const entityObj = new CoverEntity(this.hass, this.stateObj);
+    return (entityObj.isFullyClosed || entityObj.isClosing) && !assumedState;
+  }
+
+  private onOpenTap(ev): void {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:118]:
```suggestion
  private _onOpenTap(ev): void {
```

File: src/components/ha-cover-controls.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    ev.stopPropagation();
+    new CoverEntity(this.hass, this.stateObj).openCover();
+  }
+
+  private onCloseTap(ev): void {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:123]:
```suggestion
  private _onCloseTap(ev: CustomEvent): void {
```

File: src/components/ha-cover-controls.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+    ev.stopPropagation();
+    new CoverEntity(this.hass, this.stateObj).closeCover();
+  }
+
+  private onStopTap(ev): void {
```
Comment (zsarnett) [inline src/components/ha-cover-controls.ts:128]:
```suggestion
  private _onStopTap(ev: CustomEvent): void {
```

File: src/components/ha-cover-controls.ts [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+  }
+
+  private onOpenTap(ev): void {
+    ev.stopPropagation();
+    new CoverEntity(this.hass, this.stateObj).openCover();
+  }
+
+  private onCloseTap(ev): void {
+    ev.stopPropagation();
+    new CoverEntity(this.hass, this.stateObj).closeCover();
```
Comment (bramkragten) [inline src/components/ha-cover-controls.ts:125]: Can we not do this in every function, but save it on the class and update it when the state changes?

**Comment by iantrich** [comment]
Oh man...sorry guys. I meant to mark this as draft. Rushed it out before I started work this morning knowing it wasn't complete. Thanks for the reviews but sorry for wasting your time 😁

**Comment by zsarnett** [comment]
Stop WASTING my time Ian 😠 😡 🤬

File: src/components/ha-cover-controls.ts (hunk 2 above)
Comment (iantrich) [reply src/components/ha-cover-controls.ts:31]: want to maintain spacing

**Comment by iantrich** [comment]
@zsarnett I take it back. I did this 100% on purpose to spite you Zack :stuck_out_tongue_winking_eye:

File: src/components/ha-cover-controls.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+      return html``;
+    }
+
+    return html`
+      <div class="state">
+        <ha-icon-button
+          class=${classMap({
+            invisible: !this._entityObj.supportsOpen,
+          })}
+          label=${this.hass.localize("ui.dialogs.more_info_control.open_cover")}
```
Comment (iantrich) [inline src/components/ha-cover-controls.ts:44]: Am I missing something or this not supposed to produce a tooltip? I thought there was some big overhaul to support this or something?

Comment (zsarnett) [reply src/components/ha-cover-controls.ts:44]:
```suggestion
          .label=${this.hass.localize("ui.dialogs.more_info_control.open_cover")}
```

Maybe?

Comment (zsarnett) [reply src/components/ha-cover-controls.ts:44]:
```suggestion
          .title=${this.hass.localize("ui.dialogs.more_info_control.open_cover")}
```

Or is it title? idk honestly

Comment (bramkragten) [reply src/components/ha-cover-controls.ts:44]: should be `.label` but it is just for an aria label, no tooltip.

File: src/components/ha-cover-controls.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @internalProperty() private _stateObj!: HassEntity;
+
+  private _entityObj?: CoverEntity;
+
+  public set stateObj(stateObj: HassEntity) {
+    this._entityObj = new CoverEntity(this.hass, stateObj);
+    this._stateObj = stateObj;
+  }
```
Comment (bramkragten) [inline src/components/ha-cover-controls.ts:31]: Please move this to updated

File: src/components/ha-cover-controls.ts [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  private _computeCloseIcon(): string {
+    switch (this._stateObj.attributes.device_class) {
+      case "awning":
+      case "door":
+      case "gate":
+        return "hass:arrow-collapse-horizontal";
+      default:
+        return "hass:arrow-down";
+    }
+  }
```
Comment (bramkragten) [inline src/components/ha-cover-controls.ts:93]: Should we move this to <https://github.com/home-assistant/frontend/blob/dev/src/common/entity/cover_icon.ts#L4>

File: src/components/ha-cover-controls.ts [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    ev.stopPropagation();
+    this._entityObj.stopCover();
+  }
+
+  static get styles(): CSSResult {
+    return css`
+      .state {
+        white-space: nowrap;
+      }
+      .invisible {
```
Comment (bramkragten) [inline src/components/ha-cover-controls.ts:137]: 😄 
```suggestion
      .hidden {
```

**Review by bramkragten — APPROVED** [review]

**Pull request merged**